### ISSUE TITLE
FactorRange error due to mismatch between bokeh and bokehjs

### ISSUE THREAD

**Issue by rothnic** [bokeh/bokeh#3619]
```
WARNING:bokeh.core.properties:JSON had attr '_bounds_as_factors' on obj <bokeh.models.ranges.FactorRange object at 0x108140208>, which is a client-only or invalid attribute that shouldn't have been sent
```

I receive this error, but an error for `start` and `end` when using FactorRange with bokeh server. It appears that the bokehjs factorrange utilizes these properties which are not part of the python factorrange. Start and end are introduced in Range1d and FactorRange is a child of Range, which doesn't have them available. I wasn't sure why the _bounds_as_factors was being sent, but it does only appear on the bokehjs side.

**Comment by birdsarah** [comment]
@rothnic, I assume this is a server issue.

**Comment by rothnic** [comment]
This example has other issues, but this is a simplified example of what I was doing when encountering the issue:

https://gist.github.com/rothnic/18371807da60b288166f

The server get's into a loop in this example due to a different issue (i think it is separate).

**Comment by birdsarah** [comment]
@rothnic I can't get that example to produce the error you're listing

**Comment by rothnic** [comment]
@birdsarah , try again with https://gist.github.com/rothnic/18371807da60b288166f. I think you are likely hitting a different error before the one that I'm referencing that I had to patch temporarily. When you add the layout plot is checking for the axis.plot attribute and erroring out if it is set already. 

I just added a statement to clear those out before adding them, and when using the example on master, with a clean build of bokehjs, I get that error in an infinite loop.

**Comment by rothnic** [comment]
Oh, I forgot to mention that the issue will show up when you select 'c' in one of the selectors.

**Comment by mrocklin** [comment]
I've run into something similar to this.  Was there a resolution?

**Comment by bryevdv** [comment]
I can't reproduce anything with the old code updated:
```
from bokeh.models import GlyphRenderer, ColumnDataSource
from bokeh.models.plots import Plot
from bokeh.models import Circle, Range1d, LinearAxis, CategoricalAxis, Axis, FactorRange
from bokeh.layouts import row, widgetbox
from bokeh.models.widgets import Select

from bokeh.io import curdoc

data = {'a': [1, 2, 3, 4],
        'b': [3, 5, 4, 5],
        'c': ['foo', 'bar', 'foo', 'bar']}

col_names = list(data.keys())

# because dimensions and attributes of chart are modeled, this could be auto generated
x_select = Select(
        title='x', name='x_select',
        options=col_names, value='a'
)

y_select = Select(
        title='y', name='y_select',
        options=col_names, value='b'
)

x_range = Range1d(start=0, end=10)
y_range = Range1d(start=0, end=10)

x_range_cat = FactorRange(factors=data['c'])
y_range_cat = FactorRange(factors=data['c'])

x_axis = LinearAxis()
y_axis = LinearAxis()

x_axis2 = CategoricalAxis()
y_axis2 = CategoricalAxis()

plot = Plot(x_range=x_range, y_range=y_range)
plot.add_layout(x_axis, 'below')
plot.add_layout(y_axis, 'left')

ds = ColumnDataSource(data)
glyph = Circle(x='a', y='b')
renderer = GlyphRenderer(data_source=ds, glyph=glyph)

plot.renderers.append(renderer)

selectors = [x_select, y_select]

inputs = widgetbox(children=selectors)

hbox = row(children=[inputs, plot])


def update_data():

    new_renderers = []
    for renderer in plot.renderers:
        if not isinstance(renderer, Axis):
            new_renderers.append(renderer)

    plot.renderers = new_renderers

    plot.below = []
    plot.left = []

    if x_select.value == 'c':
        this_x_axis = x_axis2
        this_x_range = x_range_cat
    else:
        this_x_axis = x_axis
        this_x_range = x_range

    if y_select.value == 'c':
        this_y_axis = y_axis2
        this_y_range = y_range_cat
    else:
        this_y_axis = y_axis
        this_y_range = y_range

    this_y_axis.plot = None
    this_x_axis.plot = None

    plot.add_layout(this_x_axis, 'below')
    plot.add_layout(this_y_axis, 'left')
    plot.x_range = this_x_range
    plot.y_range = this_y_range

    glyph.x = x_select.value
    glyph.y = y_select.value


def input_change(attrname, old, new):
    update_data()

for selector in selectors:
    selector.on_change('value', input_change)

curdoc().add_root(hbox)
```

There is actually now a layout issue with adding new axes. @mattpap is this covered by another issue?

**Comment by mattpap** [comment]
@bryevdv, yes. It's likely already fixed in PR #6448.

**Comment by bryevdv** [comment]
`_bounds_as_factors` is also disappearing in #6654, closing
